CMakeLists.txt

```diff
@@ -495,6 +495,104 @@ while(library_variants_copy)
             -DLIBCXX_ENABLE_WIDE_CHARACTERS=OFF
             -DLIBUNWIND_ENABLE_THREADS=OFF
         )
+
+        #picolibc tests
+        string(REPLACE " " ";" qemu_params "${qemu_params}")
+        string(REPLACE " " ";" ld_script "${ld_script}")
+        add_custom_target(check-picolibc_${variant})
+        set(picolibc_test_ld "-T ${picolibc_SOURCE_DIR}/scripts/test-arm.ld")
+        set(qemu_command "qemu-system-arm")
+        if("${variant}" STREQUAL "aarch64")
+            set(picolibc_test_ld "")
+            set(qemu_command "qemu-system-aarch64")
+        endif()
+        string(REPLACE " " ";" picolibc_test_ld "${picolibc_test_ld}")
+        set(picolibc_tests  abort
+                    atexit
+                    complex-funcs
+                    constructor
+                    constructor-skip
+                    fenv
+                    ffs
+                    hosted-exit
+                    long_double
+                    malloc
+                    malloc_stress
+                    math_errhandling
+                    math-funcs
+                    on_exit
+                    posix-io
+                    printf_scanf
+                    printf-tests
+                    rand
+                    regex
+                    setjmp
+                    stack-smash
+                    test-efcvt
+                    test-except
+                    test-fopen
+                    test-memset
+                    test-mktemp
+                    test-put
+                    test-strchr
+                    test-strtod
+                    timegm
+                    time-sprintf
+                    time-tests
+                    tls
+                    ungetc
+                    )
+        foreach(test IN LISTS picolibc_tests)
+            add_custom_command(
+                    OUTPUT check-picolibc_${variant}_build_${test}.out
+                    COMMAND ${llvm_bin}/clang --config ${llvm_bin}/${variant}_semihost.cfg
+                    -T ../samples/ldscripts/${variant}.ld ${picolibc_test_ld}
+                    -nostdlib -lc -lg -lm -lclang_rt.builtins -lsemihost -Oz
+                    ${picolibc_SOURCE_DIR}/test/${test}.c -o check-picolibc_${variant}_build_${test}.out
+                    COMMENT "Building: ${test} for ${variant}"
+                    DEPENDS ${picolibc_SOURCE_DIR}/test/${test}.c
+            )
+            add_custom_target(
+                    check-picolibc_${variant}_${test}
+                    ${qemu_command} ${qemu_params} -semihosting -nographic -kernel check-picolibc_${variant}_build_${test}.out
+                    DEPENDS picolibc_${variant} check-picolibc_${variant}_build_${test}.out
+                    COMMENT "TEST: ${variant} ${test}"
+            )
+            add_dependencies(check-picolibc_${variant} check-picolibc_${variant}_${test})
+        endforeach()
+        add_custom_command(
```

Comment:
A lot of duplication here. Use a function like https://github.com/picolibc/picolibc/blob/main/test/CMakeLists.txt#L67